validity/url.go

```diff
@@ -21,73 +21,91 @@ import (
-	// the physical URL and the HTTP response of the resource respectively.
-	// The physical URL is typically equal to the request URL, but different
-	// in some cases; see package urlrewrite for more details.
+	// Apply returns the validity URL of a resource. physurl is the physical
+	// URL of the resource; resp is the HTTP response; vp is the period the
+	// signed exchange will be valid for. The physical URL is typically equal
+	// to the request URL but different in some cases; see package urlrewrite
+	// for more details.
 	//
 	// Note ValidityURLRule implementations can retrieve the request URL via
 	// resp.Request.URL.
-	Apply(physurl *url.URL, resp *http.Response) (*url.URL, error)
+	Apply(physurl *url.URL, resp *http.Response, vp exchange.ValidPeriod) (*url.URL, error)
 }
 
-// AppendExtDotUnixTime generates the validity URL by appending the provided
-// extension ext and the last modified time to the physical URL. For example,
+// DefaultValidityURLRule is the default rule used by webpackager.Packager.
+var DefaultValidityURLRule ValidityURLRule = AppendExtDotLastModified(".validity")
+
+// AppendExtDotLastModified generates the validity URL by appending ext
+// and the resource's last modified time. For example:
 //
 //     https://example.com/index.html
 //
-// may yield the following validity URL:
+// would receive a validity URL that looks like:
 //
 //     https://example.com/index.html.validity.1561984496
 //
-// ext usually starts with a dot ("."); AppendExtDotUnixTime does not insert
-// the one automatically. ext is therefore ".validity" rather than "validity"
+// ext usually starts with a dot ("."). AppendExtDotExchangeDate does not
+// insert it automatically. ext is thus ".validity" rather than "validity"
 // in the example above.
 //
-// The last modified time is represented by a UNIX timestamp, and taken from
-// the Last-Modified header of the response. When Last-Modified is missing or
-// has an unparsable value, AppendExtDotUnixTime uses the now argument instead.
-//
-// AppendExtDotUnixTime does not expect the physical URL to contain Query or
-// Fragment. They will be stripped off from the validity URL.
+// The last modified time is taken from the Last-Modified header field in
+// the HTTP response and represented in UNIX time. If the Last-Modified
+// is missing or unparsable, AppendExtDotLastModified uses the date value
+// of the signed exchange signature (vp.Date).
 //
-// Apply returns an error when physurl looks like a directory, e.g. the path
-// ends with a slash.
-func AppendExtDotUnixTime(ext string, now time.Time) ValidityURLRule {
-	return &appendExtDotUnixTime{ext, now}
+// The returned ValidityURLRule does not accept physurl that looks like
+// a directory (e.g. has Path ending with a slash): Apply returns error
+// for such physurl. Note it does not occur if urlrewrite.IndexRule is
+// used. Also the rule does not expect physurl to have Query or Fragment:
```

Comment:
"does not expect" sounds a bit unclear (there's another, "does not accept" as well.)
Honestly I have no idea about the ideal treatment of those with Query or Fragment, but in order to clarify the behavior, especially the domain of this function, either

- explicitly reject URLs with nonempty Query or Fragment (make it an error or just a precondition could be another problem though)
- accepts them and explain the current behavior, "append to the Path component, while leaving Query and Fragment as they are" and (optionally) adding a concern about it

would be more helpful.

Reply:
Reworded godoc.

I agree that the way Query is currently handled is not necessarily ideal, but I don't intend to change the behavior with this pull request. Fragment should be dropped anyway: browsers drop it when sending an HTTP request.